### ISSUE TITLE
C# 7.x §15.15.1 interface, delegate, or enum as task type

### ISSUE THREAD

**Issue by KalleOlaviNiemitalo** [dotnet/csharpstandard#859]
**Describe the bug**

In the C# 7.x draft, §15.15.1 (Async functions / General) says:

> A task type shall be a class or struct type that is associated with a ***task builder type*** via the attribute `System.Runtime.CompilerServices.AsyncMethodBuilderAttribute`. Such types are referred to in this specification as `«TaskBuilderType»<T>` and `«TaskBuilderType»`.

However, Roslyn allows a task type that is neither a class type nor a struct type, as long as it has the AsyncMethodBuilderAttribute.

**Example**

An interface should be allowed as a task type.

```csharp
using System;
using System.Runtime.CompilerServices;

class C {
    async ICustomTask M() {}
}

[AsyncMethodBuilder(typeof(CustomTaskBuilder))]
public interface ICustomTask {}

public struct CustomTaskBuilder {
    public static CustomTaskBuilder Create() => default;
    public void Start<TStateMachine>(
        ref TStateMachine stateMachine)
        where TStateMachine : IAsyncStateMachine {}
    public void SetStateMachine(
        IAsyncStateMachine stateMachine) {}
    public void SetException(
        Exception exception) {}
    public void SetResult() {}
    public void AwaitOnCompleted<TAwaiter,TStateMachine>(
        ref TAwaiter awaiter,
        ref TStateMachine stateMachine)
        where TAwaiter : INotifyCompletion
        where TStateMachine : IAsyncStateMachine {}
    public void AwaitUnsafeOnCompleted<TAwaiter,TStateMachine>(
        ref TAwaiter awaiter,
        ref TStateMachine stateMachine)
        where TAwaiter : INotifyCompletion
        where TStateMachine : IAsyncStateMachine {}
    public ICustomTask Task => default;
}
```

**Expected behavior**

In §15.15.1, also allow an interface, delegate, or enum type as a task type.  Say just "A task type shall be a type that" rather than "A task type shall be a class or struct type that" -- except <https://github.com/dotnet/csharpstandard/issues/854> will need more changes here.

**Additional context**

The C# 7.0 [Async Task Types in C#](https://github.com/dotnet/csharplang/blob/05ed9dcfba84a038e02301a9b267516edfe14159/proposals/csharp-7.0/task-types.md) proposal does say: "A task type is a `class` or `struct` with an associated *builder type* identified with `System.Runtime.CompilerServices.AsyncMethodBuilderAttribute`."

<https://github.com/dotnet/corefx/pull/11140> initially added `internal sealed class AsyncBuilderAttribute` with `AttributeTargets.Class | AttributeTargets.Struct | AttributeTargets.Interface`.

<https://github.com/dotnet/corefx/pull/11505> renamed it to `public sealed class AsyncMethodBuilderAttribute`, and added `AttributeTargets.Delegate | AttributeTargets.Enum`.  There was some discussion about those attribute targets.

In §C.3 (Standard Library Types not defined in ISO/IEC 23271), AsyncMethodBuilderAttribute has `AttributeTargets.Class | AttributeTargets.Struct | AttributeTargets.Interface | AttributeTargets.Delegate | AttributeTargets.Enum | AttributeTargets.Method`, but see <https://github.com/dotnet/csharpstandard/issues/858> regarding `AttributeTargets.Method`.

**Comment by jskeet** [comment]
I'm happy to include an interface here, as that genuinely makes sense - but being able to treat enums and delegate types as task types feels like a quirk of Roslyn that we shouldn't propagate into the standard.

@BillWagner: Initial thoughts?

(I'll try to come up with a PR for this and #854 in the same paragraph.)

**Comment by BillWagner** [comment]
I agree with @jskeet that this feels like an implementation quirk.

Let me ask a couple folks for additional thoughts.

**Comment by BillWagner** [comment]
@jskeet 

The consensus is that both enums and delegates don't have a useful purpose and shouldn't be part of the standard. It's an implementation quirk.

**Comment by jskeet** [comment]
@BillWagner: Cool. In that case, I believe #875 is ready to merge, if you're happy with it.

**Comment by jskeet** [comment]
Fixed in #875.
